site/app/libraries/Access.php

```diff
@@ -458,4 +594,121 @@ private function resolveNewGradeable($g) {
+        foreach (explode(DIRECTORY_SEPARATOR, $file) as $part) {
+            if ($part == ".." || $part == ".") {
+                return false;
+            }
+        }
+
+        //Make sure it starts with the dir base
+        if (!Utils::startsWith($file, $info["base"])) {
+            //This both prevents people from accessing files outside the base dir
+            // and lets us have relative paths. Convenient!
+            if ($file[0] === "/") {
+                $file = substr($file, 1);
+            }
+
+            $relative_path = $file;
+            $file = $info["base"] . "/" . $file;
+        } else {
+            $relative_path = substr($file, strlen($info["base"]) + 1);
+        }
+
+        //If it doesn't exist we can't access it
+        if (!file_exists($file)) {
+            return false;
+        }
+
+        //Check if the relative path starts with the right directory
+        $subpart_types = $info["subparts"];
+        $subpart_values = explode("/", $relative_path);
+
+        //Missing necessary directory path
+        if (count($subpart_values) <= count($subpart_types)) {
+            return false;
+        }
+        $end_path = implode("/", array_slice($subpart_values, count($subpart_types)));
+        $subpart_values = array_slice($subpart_values, 0, count($subpart_types));
+        $subpart_types[] = "path";
+        $subpart_values[] = $end_path;
+
+        //To array of [type, value]
+        $subparts = array_combine($subpart_types, $subpart_values);
+
+        //So we can extract parameters from the path
+        foreach ($subpart_types as $type) {
+            $value = $subparts[$type];
+            switch ($type) {
+                case "gradeable":
+                    //If we already have a gradeable in the args, make sure this file actually belongs to it
+                    if (array_key_exists("gradeable", $args)) {
+                        //Check if the gradeable matches
+
+                        /* @var Gradeable|null $gradeable */
+                        /* @var \app\models\gradeable\Gradeable|null $new_gradeable */
+                        list($gradeable, , $new_gradeable,) = $this->resolveNewGradeable($args["gradeable"]);
+                        if (($new_gradeable ?? $gradeable)->getId() !== $value) {
+                            return false;
+                        }
+                    }
+                    $args["gradeable"] = $this->core->getQueries()->getGradeableConfig($value);
+                    break;
+                case "submitter":
+                    $submitter = $this->core->getQueries()->getSubmitterById($value);
+                    if ($submitter !== null) {
+                        $args["submitter"] = $submitter;
+                        if (array_key_exists("gradeable", $subparts)) {
+                            //If we already have a graded gradeable in the args, make sure this file
+                            // actually belongs to it
```

Comment:
Does this comment mean that we check that the file is a submission file / results file and we need to check that the student is allowed to access it?  I can't figure out where that might be happening below.

Reply:
The thought was that you could pass both a gradeable and a file to `canI`, and this would make sure that file was associated with that gradeable. Not sure how necessary it is but it might prevent some weirdness if you try to access a file across gradeables.